File: python/array/palindrome.py
# Given a string s, return the longest palindromic substring in s.

# Example 1:

# Input: s = "babad"
# Output: "bab"
# Explanation: "aba" is also a valid answer.
# Example 2:

# Input: s = "cbbd"
# Output: "bb"

'''
examples = [
    "radar",
    "A man a plan a canal Panama",
    "race a car",
    "Madam Im Adam",
    "noon",
    "level"
]
'''

def build_center_expansion(s, left, right, center):
    while left >= 0 and right < len(s):
        if s[left] == s[right]:
            center = s[left] + center + s[right]
            left -= 1
            right += 1
        else:
            break
    return center

class Solution:
    def longestPalindrome(self, s: str) -> str:
        # Find the longest palindrome with center expansion
        longest = ""
        # Center expansion with odd length
        for i in range(len(s)):
            left = i - 1
            right = i + 1
            center = s[i]
            tmp = build_center_expansion(s, left, right, center)
            if len(tmp) > len(longest):
                longest = tmp
                
        # Center expansion with even length, empty center
        for i in range(len(s)-1):
            left = i
            right = i + 1
            center = ""
            tmp = build_center_expansion(s, left, right, center)
            if len(tmp) > len(longest):
                longest = tmp
                
        return longest

File: typescript/src/string/palindrome.ts
function buildCenterExpansion(s: string, left: number, right: number): string {
    while (left >= 0 && right < s.length && s[left] === s[right]) {
        left--;
        right++;
    }
    return s.slice(left + 1, right);
}

export function longestPalindrome(s: string): string {
    let longest = '';
    for (let i = 0; i < s.length; i++) {
        const odd = buildCenterExpansion(s, i, i);
        const even = buildCenterExpansion(s, i, i + 1);
        longest = longest.length > odd.length ? longest : odd;
        longest = longest.length > even.length ? longest : even;
    }
    return longest;
}
